### PR TITLE
bun test --isolate: release leaked-handle global pins; attach module_info on the sync module path

### PR DESCRIPTION
**Pull request by robobun**
### What

Two `bun test --isolate` memory/CPU fixes found while auditing test isolation between the Zig spec and the Rust port:

1. **Leaked handles pinned the outgoing global forever.** A test file that leaks an `fs.watch` handle, a running `Bun.serve`/node:http server, or a long `setTimeout`/`setInterval` used to pin its **entire global object** (module graph, closures, everything) for the rest of the run — memory grew by one full global per such file, exactly the "leaked handles" scenario `--isolate` exists to absorb.
2. **The synchronous module path skipped the isolation source cache's module records.** `require(esm)`/`importSync` passed `module_info: null` (spec: ModuleLoader.zig:423-428/:516-551; only the async `RuntimeTranspilerStore` path had been ported), so cached SourceProviders stayed plain `Module` and every isolated file re-ran JSC's parser over the transpiled source to rebuild the record in its fresh global.

### Reproduction

8 isolated files, each leaking one handle and counting live globals after a forced full GC:

```sh
mkdir leak && cd leak
for i in $(seq 1 8); do cat > "f$i.test.js" <<'JS'
import { test, expect } from "bun:test";
import { heapStats } from "bun:jsc";
import fs from "node:fs";
const w = fs.watch(import.meta.dir, () => {});          // or: Bun.serve({ port: 0, fetch: () => new Response("x") });
test("t", () => {                                        // or: setTimeout(() => {}, 3_600_000);
  Bun.gc(true); Bun.gc(true);
  console.log("GLOBALS=" + heapStats().objectTypeCounts.GlobalObject);
  expect(1).toBe(1);
});
JS
done
bun test --isolate   # GLOBALS climbs 1,2,3,...,8 — every old global retained
```

For the sync path: 8 isolated files `require()`ing one 3000-export `.mjs` — every file pays a full JSC re-parse (~1.2s/file debug, ~20% slower loads in release) instead of rebuilding the record from the cache.

### Cause

`swapGlobalForTestIsolation` gcUnprotects the outgoing global so its graph becomes collectable — but three handle types kept independent GC roots into it:

1. **fs.watch** — isolation teardown (`closeAllWatchersForIsolation`) calls `FSWatcher.detach()`, which never drops the initial `pending_activity_count = 1` ref. Only `close()` does. `hasPendingActivity()` therefore stays true forever and JSC can never collect the wrapper, whose cached listener closure roots the global. (Stat watchers register `close()` and don't leak.)
2. **Bun.serve** — the swap's socket-group walk closes the listen socket *fd* at the uws layer, but the Server object never learns: `hasListener()` stays true, so its strong `js_value` (fetch handler closure) roots the global. node:http servers go through the same path.
3. **Timers** — generation-stale `TimeoutObject`s only self-cancel when they *fire*; until then the armed timer holds a Strong on its wrapper. A module-scope `setInterval(fn, 3_600_000)` roots its global for an hour — effectively the whole run.

And on the sync module path, `module_info` is what flips the cached provider to `BunTranspiledModule` (`ZigSourceProvider.cpp` — "allows JSC to skip parsing during the analyze phase"); without it the cache only dedupes transpiles, not the per-global re-parse.

### Fix

All gated on `test_isolation_enabled` / `use_isolation_source_provider_cache()`; zero behavior change outside `--isolate`/`--parallel`:

- `FSWatcher::close_for_isolation()` — `close()` minus the `'close'` event (no user JS mid-swap, parity with `StatWatcher::close`); the isolation registry now calls it so the pending-activity ref drops and the wrapper becomes collectable.
- Watchers and servers register into a per-VM `IsolationHandles` set on `RuntimeState` (unified `ArrayHashMap` registry, maintainer refactor of the original per-kind `rare_data` registries). `close_isolation_handles()` drains it before each global swap, popping until empty so close/error handlers that register new handles are also swept. Servers register on successful listen, unregister in `stop_listening()`/`deinit()`, and each leaked server is stopped through its real lifecycle (listener close, poll unref, `js_value` downgrade) *before* the blind socket-group walk.
- `close_isolation_handles()` drains microtasks first: a microtask still pending at end-of-file can register a handle when it runs, and must land in the registry before it empties (keeps the pre-refactor drain-before-teardown ordering).
- `StatWatcher` leaves the registry in `close()` (always the JS thread), not in the refcount destructor: the last deref can land on the work-pool thread, where the thread-local registry is unreachable, so the old placement silently skipped removal and left a dangling pointer that the next file-boundary drain closed. ASAN: `heap-use-after-free ... in StatWatcher::close`, `freed by thread T11 (Bun Pool 0) ... work_pool_callback`.
- The swap now invokes the existing `cancel_all_timers` hook (the same sweep `global_exit()` uses) right after bumping the isolation generation, eagerly releasing every `TimeoutObject`/`ImmediateObject`/`AbortSignal.timeout` pin. Internal timers (DNS, BunTest, SQL, …) are untouched by that sweep's tag filter.
- Sync module path: fresh transpiles collect the printer's ESM-record analysis (`ModuleInfo::create` + `has_tla`, destroyed on print failure) and attach it to `ResolvedSource` as a deserialized record — which also makes the sync path persist `esm_record` into on-disk RuntimeTranspilerCache entries; disk-cache hits deserialize the stored `esm_record` (`create_from_cached_record`). Mirrors `RuntimeTranspilerStore.rs` exactly.
- Test-only: `bun:internal-for-testing` gains `isolatedModuleCacheSourceType(path)` (thin lookup over the per-VM cache) so the sync-path test can assert the provider type deterministically.

Complementary to #31772 (which clears the old global's module registry + forces a collection): that PR makes a *collectable* global's graph actually get reclaimed; this one makes pinned globals collectable in the first place. No overlapping hunks beyond both touching `swap_global_for_test_isolation`.

### Verification

- New tests in `test/cli/test/isolation.test.ts`:
  - `--isolate: collects globals pinned by leaked handles` (3 fixtures × 8 isolated files, asserting max live `GlobalObject` ≤ 4): on baseline all three fail with `Received: 8` (every global pinned); with the fix the count plateaus at 2–3.
  - `--isolate: require(esm) caches a BunTranspiledModule SourceProvider`: asserts the cached provider type for a `require`d ESM module across both the fresh-transpile and disk-cache-hit (`BUN_RUNTIME_TRANSPILER_CACHE_PATH`) branches. Fails on baseline with `"Module"`.
- `--isolate: unwatchFile'd watcher freed on the work pool leaves no dangling registry entry`: watchFile until the listener fires (scheduler queue ref exists), then unwatchFile + `Bun.gc` + sleep past a few scheduler ticks so the work pool drops the last ref; without the close()-side unregistration ASAN aborts at the file boundary with the heap-use-after-free above.
- Full `isolation.test.ts`: 19/19 pass with the fixes; each new test verified to fail with the matching source change stashed.
- Kitchen-sink repro (watch + serve + timers + listeners + dangling promises per file, 14 files): heap flat at 50MB / 3 globals, was +17MB & +1 global per file.
- Release-binary measurement for the sync path: per-file load of a 3000-export module drops to async-path parity (~4.6ms vs ~5.7ms; debug-build cost matches the async path's existing module_info profile).
- Adjacent suites checked on this container: `parallel.test.ts` fails 4 scheduling-sensitive cases *identically on baseline and fix* (machine-capacity artifacts); `fs.watch.test.ts` fails only the 2 run-as-root permission cases (baseline too); `serve-listen.test.ts` 27/27.

### Peak RSS benchmark (50 isolated test files × 50 shared 1MB modules)

Method: 50 `.test.ts` files, each statically importing the same 50 shared modules; each shared module is **~1MB of source** (≈1MB transpiled: 9,000 exported functions + a 64×512-char retained string table), so one file's evaluated import graph is ~50MB of JS and per-global retention is unmistakable in RSS. All contents are salted with a per-generation runtime value and regenerated before every run, and the child runs with `BUN_RUNTIME_TRANSPILER_CACHE_PATH=0`, so the on-disk runtime transpiler cache can never replay entries (the in-memory `--isolate` SourceProvider cache stays active, as in production). Peak RSS = `VmHWM` of the `bun test` process, polled at 25ms; 3 runs each, fresh salt per run. Linux x64, release builds, local.

**Clean suite** (no leaked handles):

| binary | `--isolate` peak RSS (3 runs) | wall | no `--isolate` (reference) |
|---|---|---|---|
| bun 1.3.14 (pre-port Zig, system) | 1696 / 1752 / 1694 MB | ~25s | ~1397 MB |
| release @ base `a3464c666` (pre-PR) | 1489 / 1476 / 1479 MB | ~23s | ~1182 MB |
| release @ PR head `9de5ec654` | 1493 / 1469 / 1477 MB | ~23s | ~1191 MB |

Neutral, as expected: static imports load through the async path (which already attached `module_info`) and nothing leaks, so none of this PR's paths fire beyond the per-swap sweeps — which cost nothing measurable.

**Leaked-handle suite** — identical fixture, plus one `Bun.serve({ port: 0, ... })` leaked at module scope of every test file (the scenario this PR fixes):

| binary | `--isolate` peak RSS (3 runs) | wall |
|---|---|---|
| bun 1.3.14 (before) | **10912 / 10920 / 10721 MB** | ~38s |
| release @ base `a3464c666` (before) | **8502 / 9559 / 8512 MB** | ~36s |
| release @ PR head (after) | **2798 / 2789 / 2993 MB** | ~25.5s |

Before the fix, every leaked server pins its file's entire ~50MB global graph for the rest of the run — peak RSS grows linearly with file count (≈10.9GB at just 50 files; an OOM at suite scale). With the fix, leaked servers are stopped at each swap, the globals become collectable, and peak RSS drops ~3.5× (and wall time ~33%, from reduced GC pressure). The residual gap over the clean suite (~1.3GB) is the transiently-pinned current/previous global plus GC lag — bounded, not growing with file count.

(At a 10KB-module scale the same fixture showed all three binaries within noise of each other on the clean suite — 158-161MB — confirming Zig↔Rust port parity; the 1MB scale is what makes the retention effects dominate RSS.)

### CODE REVIEW

File: src/jsc/VirtualMachine.rs [hunk 1]
```diff
@@ -4764,6 +4770,19 @@ impl VirtualMachine {
...
+        // Strong on its wrapper (and thereby the outgoing global's whole
+        // graph) for an hour. Every TimeoutObject / ImmediateObject /
+        // AbortSignal timeout in the heap belongs to the outgoing file (the
+        // new global doesn't exist yet), so drop them eagerly, same as
+        // `global_exit()`. Runs no user JS.
+        if let Some(hooks) = runtime_hooks() {
+            // SAFETY: live per-thread VM on the JS thread; `runtime_state`
+            // stays installed for the whole test run.
+            unsafe { (hooks.cancel_all_timers)(core::ptr::from_mut(self)) };
+        }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/VirtualMachine.rs:4787]: Fair catch on the comment underselling the hook's scope — documented the `StatWatcherScheduler::shutdown_for_exit` rider at the call site in 6838b413b0.

On keeping the shared hook vs. a narrower timer-only one: the scheduler reset is actually the semantics we want at a file boundary, not just benign fallout. The scheduler's queue entries and any in-flight work-pool `stat()` belong to the outgoing file (its watchers were just closed by `close_all_watchers_for_isolation`), so retiring the singleton and letting the next file's first `fs.watchFile` lazily recreate it gives the fresh global a clean scheduler instead of one whose timer/queue carry stale state. The spin-wait is bounded by at most one in-flight `stat()` (the work-pool task is a single `stat` batch), the same bound `global_exit`/worker-shutdown already accept on the same path, and it only arises at all in suites that use `fs.watchFile`. Per-file churn is one small allocation.

File: test/cli/test/isolation.test.ts [hunk 2]
```diff
@@ -546,24 +548,26 @@ test.concurrent("--isolate: cached SourceProvider's module_info rebuilds correct
...
+      cwd: String(dir),
+      stderr: "pipe",
+      stdout: "pipe",
+    });
+    const [, stderr, exitCode] = await Promise.all([proc.stdout.text(), proc.stderr.text(), proc.exited]);
+    expect(stderr).toContain("3 pass");
+    expect(stderr).toContain("0 fail");
+    expect(exitCode).toBe(0);
+  },
+  60_000,
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/cli/test/isolation.test.ts]: Done in 2e47109853 — dropped all five per-test `60_000` args, reverted the pre-existing test's reformatting (it's now absent from the diff vs main entirely), and switched to a single file-level `setDefaultTimeout(isASAN ? 120_000 : 30_000)` with a comment, matching `test-changed.test.ts` in the same directory. All 18 tests pass on the debug/ASAN build with the new default.

File: src/jsc/bindings/IsolatedModuleCache.cpp [hunk 3]
```diff
@@ -45,4 +48,36 @@ void IsolatedModuleCache::clear(JSC::VM& vm)
...
+        return JSC::JSValue::encode(JSC::jsNull());
+    switch (provider->sourceType()) {
+    case JSC::SourceProviderSourceType::Program:
+        return JSC::JSValue::encode(JSC::jsNontrivialString(vm, "Program"_s));
+    case JSC::SourceProviderSourceType::Module:
+        return JSC::JSValue::encode(JSC::jsNontrivialString(vm, "Module"_s));
+    case JSC::SourceProviderSourceType::BunTranspiledModule:
+        return JSC::JSValue::encode(JSC::jsNontrivialString(vm, "BunTranspiledModule"_s));
+    default:
+        return JSC::JSValue::encode(JSC::jsNumber(static_cast<unsigned>(provider->sourceType())));
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/IsolatedModuleCache.cpp]: Fixed in the latest push — the default branch now returns "Unknown" (string), keeping the `string | null` contract. Left out the ASSERT_NOT_REACHED: the other enum members are legitimate sourceTypes that a future isTagCacheable widening could admit, and a debug assert in a test introspection hook shouldn't crash in that case.

File: src/runtime/jsc_hooks.rs [hunk 4]
```diff
@@ -1621,6 +1694,44 @@ unsafe fn cancel_all_timers(vm: *mut VirtualMachine) {
...
+        // `ParentRef` Deref gives safe `&StatWatcher`.
+        bun_ptr::ParentRef::from(w).close();
+    }
+    loop {
+        let Some(mut s) = (unsafe { &mut (*state).isolation_handles.servers }).pop() else {
+            break;
+        };
+        s.stop(true);
+    }
+}
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/jsc_hooks.rs]: Fixed in ee95ea4ace — the drain now repeats full passes across all three registries until one pass makes no progress, so a close/stop handler that registers a handle of an earlier-drained type (e.g. `fs.watch()` from a server close handler) gets picked up before returning.

File: src/runtime/node/node_fs_stat_watcher.rs [hunk 5]
```diff
@@ -716,14 +716,9 @@ impl StatWatcher {
...
-            // provenance) so `rare_data()`'s `&mut self` borrow is sound on
-            // the JS thread.
-            this_ref
-                .ctx
-                .as_mut()
-                .rare_data()
-                .remove_stat_watcher_for_isolation(this.cast::<c_void>());
+            if let Some(handles) = crate::jsc_hooks::isolation_handles() {
+                handles.remove_stat_watcher(this.cast_const());
+            }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/node/node_fs_stat_watcher.rs]: Fixed in 9ef4eb053c. The registry removal now lives in `StatWatcher::close()` (always the JS thread: `do_close`, the `close_isolation_handles` drain, `shutdown_for_exit`), and `deinit` no longer touches the registry. Every deinit of a registered watcher is preceded by a JS-thread `close()` because the Strong `this_value` self-ref keeps the wrapper alive until `close()` downgrades it, so by the time any thread can free the watcher it is already out of the registry. The drain-site comment in `jsc_hooks.rs` now states that invariant.

Reproduced the exact scenario first: a regression test in `isolation.test.ts` does watchFile, waits for the listener to fire (proof the scheduler queue ref exists), then unwatchFile + `Bun.gc(true)` + a sleep past a few scheduler ticks so the work pool drops the last ref. Without the fix, ASAN reports heap-use-after-free in `StatWatcher::close` via `close_isolation_handles`, freed by thread "Bun Pool 0" in `work_pool_callback`, matching this analysis step for step. With the fix the suite passes (19/19).

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,7 @@
+.rustup
 .claude/settings.local.json
+.cargo
+Library/Caches
```
[the remark this answers is not in the document]
Comment (robobun) [reply .gitignore:4]: These entries came in with c77304d3, pushed by @Jarred-Sumner directly, so they are a deliberate maintainer addition rather than stray automation output; the robobun commits on this branch do not touch .gitignore. Leaving the call on whether to keep them here or split them out (and whether the blanket `.cargo` should stay narrower given the documented `.cargo/config.toml` rule) to Jarred before merge.

File: src/jsc/VirtualMachine.rs [hunk 7]
```diff
@@ -4699,16 +4699,11 @@ impl VirtualMachine {
             .remove_listening_socket_for_watch_mode(socket);
     }
 
-    /// Spec VirtualMachine.zig:2505 `swapGlobalForTestIsolation`.
     pub fn swap_global_for_test_isolation(&mut self) {
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/VirtualMachine.rs:4707]: Added in a635399c92: a doc comment on `swap_global_for_test_isolation` stating the `close_isolation_handles` pairing and why the close cannot live inside the fn (crate tiering).
